File: Java/LeetCode/PrimeNumberOfSetBitsInBinaryRepresentation.java
// https://leetcode.com/problems/prime-number-of-set-bits-in-binary-representation/description
// 762. Prime Number of Set Bits in Binary Representation
// EASY
/* Given two integers left and right, return the count of numbers in the inclusive range [left, right] having a prime number of set bits in their binary representation.
Recall that the number of set bits an integer has is the number of 1's present when written in binary.
For example, 21 written in binary is 10101, which has 3 set bits.

Example 1:
Input: left = 6, right = 10
Output: 4
Explanation:
6  -> 110 (2 set bits, 2 is prime)
7  -> 111 (3 set bits, 3 is prime)
8  -> 1000 (1 set bit, 1 is not prime)
9  -> 1001 (2 set bits, 2 is prime)
10 -> 1010 (2 set bits, 2 is prime)
4 numbers have a prime number of set bits.
Example 2:
Input: left = 10, right = 15
Output: 5
Explanation:
10 -> 1010 (2 set bits, 2 is prime)
11 -> 1011 (3 set bits, 3 is prime)
12 -> 1100 (2 set bits, 2 is prime)
13 -> 1101 (3 set bits, 3 is prime)
14 -> 1110 (3 set bits, 3 is prime)
15 -> 1111 (4 set bits, 4 is not prime)
5 numbers have a prime number of set bits.
Constraints:
1 <= left <= right <= 10^6
0 <= right - left <= 10^4 */

class PrimeNumberOfSetBitsInBinaryRepresentation {
    // We only need primes up to 19 because R≤10^6<2^20 -> 20 bits
    //private static final Set<Integer> PRIMES = Set.of(2, 3, 5, 7, 11, 13, 17, 19);
    private static final int[] PRIMES = new int[] {0,0,1,1,0,1,0,1,0,0,0,1,0,1,0,0,0,1,0,1,0};

    public int countPrimeSetBits(int left, int right) {
        int ans = 0;
        for (int x = left; x <= right; ++x)
        // isPrime (a common method) and getSetBit (instead of bitCount) can be used
            //if (PRIMES.contains(Integer.bitCount(x))) {
            if (PRIMES[Integer.bitCount(x)] == 1) {
                ans++;
            }
        return ans;
    }

    static boolean isPrime(int n) {
        // Corner case
        if (n <= 1)
            return false;
        // Check from 2 to sqrt(n)
        for (int i = 2; i <= Math.sqrt(n); i++)
            if (n % i == 0) {
                return false;
            }
        return true;
    }

    static int getSetBit(int n){
        int ans = 0;
        while (n > 0){
            if (n % 2 != 0) ans++;
            n = n / 2;
        }
        return ans;
    }
}

File: Kotlin/LeetCode/PrimeNumberOfSetBitsInBinaryRepresentation.kt
class Solution {
    // We only need primes up to 19 because R≤10^6<2^20 -> 20 bits
    // private val PRIMES = setOf(2, 3, 5, 7, 11, 13, 17, 19)
    private val PRIMES = intArrayOf(0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0)

    fun countPrimeSetBits(left: Int, right: Int): Int {
        var ans = 0
        for (x in left..right) {
            // isPrime (a common method) and getSetBit (instead of countOneBits) can be used
            // if (PRIMES.contains(x.countOneBits())) {
            if (PRIMES[x.countOneBits()] == 1) {
                ans++
            }
        }
        return ans
    }

    fun isPrime(n: Int): Boolean {
        // Corner case
        if (n <= 1) return false
        // Check from 2 to sqrt(n)
        val limit = kotlin.math.sqrt(n.toDouble()).toInt()
        for (i in 2..limit) {
            if (n % i == 0) {
                return false
            }
        }
        return true
    }

    fun getSetBit(n: Int): Int {
        var count = 0
        var num = n
        while (num > 0) {
            if (num % 2 != 0) count++
            num /= 2
        }
        return count
    }
}
